### ISSUE TITLE
iAvg function does not calculate arithmetic mean

### ISSUE THREAD

**Issue by mainzelM** [Raruto/leaflet-elevation#187]
### Subject of the issue
The iAvg function at https://github.com/Raruto/leaflet-elevation/blob/2bbaa3dd271087e77685dc018730bf7769f6514b/src/utils.js#L136 which is commonly called in the handlers does not calculate the arithmetic mean of the provides values (but something else I cannot describe). For example for a series of four values 100, 100, 200, 200, I'd expect the average to be 150 but actually it will be 168.75:

```js
iAvg(100, 0);     => 50     // expected average for [100] is 100
iAvg(100, 50);    => 75     // expected average for [100, 100] is 100
iAvg(200, 75);    => 137.5  // expected average for [100, 100, 200] is 133.3
iAvg(200, 137.5); => 168.75 // expected average for [100, 100, 200, 200] is 150
```
However, the avg function alone cannot fix this problem, as it does not know the number of values that made up the previous avg value.

### Your environment
* leaflet-elevation: 2.1.0

**Comment by Raruto** [comment]
Hi mainzeIM,
thanks for the report, probably this conceptual error is not related only to version 2.x (even in the previous ones it should be calculated in a similar way), this is more or less the same code as in version 1.8.0:

https://github.com/Raruto/leaflet-elevation/blob/7955321b69810b57d8e7b29838594b45f239add2/src/acceleration.js#L46-L85

Anyway, I think a neat way to solve this question would be to use [generator functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*) (as they are already used within: [https://github.com/tmcw/togeojson](https://github.com/placemark/togeojson)). But honestly, I don't have a lot of experience with it... (if you want to deepen or think you already have a solution I would be grateful if you wanted to share it with everyone).

_Have a nice day,
Raruto_

**Comment by mainzelM** [comment]
I don't know generator functions either but if you want to stick to the existing protocol, you can store the number of previous values in a closure:
```
    > function Avg() {
        let count = 0;
        return function(next, prevAvg) {
            let sumSoFar = next + ((prevAvg ?? 0) * count);
            return sumSoFar / ++count;
        }
    }
   > const iAvg = Avg();
   > iAvg(200, iAvg(200, iAvg(100, iAvg(100, undefined))));
   150
```

**Comment by Raruto** [comment]
> I don't know generator functions either but if you want to stick to the existing protocol, you can store the number of previous values in a closure:

Broadly speaking generators are more or less the same thing:

```js
function* infinite() {
    let index = 0;

    while (true) {
        yield index++;
    }
}

const generator = infinite(); // "Generator { }"

console.log(generator.next().value); // 0
console.log(generator.next().value); // 1
console.log(generator.next().value); // 2
// ...
```

so probably, to keep things consistent, all [these functions](https://github.com/Raruto/leaflet-elevation/blob/ee93c5ae4b6668650e5950fbb69c5831cc18b320/src/utils.js#L131-L137) will need to return some sort of closure too?

```js
const iAvg = Avg();
const iMax = Max();
const iMin = Min();
```

**Comment by mainzelM** [comment]
> > I don't know generator functions either but if you want to stick to the existing protocol, you can store the number of previous values in a closure:
> 
> Broadly speaking [generators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator) are more or less the same thing:

Yes, but as it seems, passing in a value and receiving it at the same time is [not possible](https://bytethisstore.com/articles/pg/js-generator-function) for the first value:

```
function* avgGenerator() {
    let current_avg = yield 0;
    let count = 1;
    while (true) {
        const n = yield current_avg;
        current_avg = (n + (current_avg * count))/++count;
    }
}
const avg = avgGenerator();

avg.next(100).value;
0
avg.next(100).value;
100
avg.next(200).value;
150
avg.next(200).value;
166.66666666666666
```

> so probably, to keep things consistent, all [these functions](https://github.com/Raruto/leaflet-elevation/blob/ee93c5ae4b6668650e5950fbb69c5831cc18b320/src/utils.js#L131-L137) will need to return some sort of closure too?
> 
> ```js
> const iAvg = Avg();
> const iMax = Max();
> const iMin = Min();
> const iSum = Sum();
> ```

Well, yes, that would probably make sense and also allow you to keep additional state, e.g. to allow max() to compare the current value with the previous value and ignore the current value if it is too far above the previous value (because of a sensor error).

**Comment by mainzelM** [comment]
:thumbsup:, thanks a lot!

**Comment by Raruto** [comment]
Patch released in version **2.1.1**
